public/app/features/dashboard/state/runRequest.ts

```diff
@@ -0,0 +1,154 @@
+ */
+export function processResponsePacket(packet: DataQueryResponse, state: RunningQueryState): RunningQueryState {
+  const request = state.panelData.request;
+  const packets: MapOfResponsePackets = {
+    ...state.packets,
+  };
+
+  packets[packet.key || 'A'] = packet;
+
+  // Update the time range
+  let timeRange = request.range;
+  if (isString(timeRange.raw.from)) {
+    timeRange = {
+      from: dateMath.parse(timeRange.raw.from, false),
+      to: dateMath.parse(timeRange.raw.to, true),
+      raw: timeRange.raw,
+    };
+  }
+
+  const combinedData = flatten(
+    lodashMap(packets, (packet: DataQueryResponse) => {
+      return packet.data;
+    })
+  );
+
+  const panelData = {
+    state: LoadingState.Done,
+    series: combinedData,
+    request: {
+      ...request,
+      range: timeRange,
+    },
+  };
+
+  return { packets, panelData };
+}
+
+/**
+ * This function handles the excecution of requests & and processes the single or multiple response packets into
+ * a combined PanelData response.
+ * It will
+ *  * Merge multiple responses into a single DataFrame array based on the packet key
+ *  * Will emit a loading state if no response after 50ms
+ *  * Cancel any still runnning network requests on unsubscribe (using request.requestId)
+ */
+export function runRequest(datasource: DataSourceApi, request: DataQueryRequest): Observable<PanelData> {
+  let state: RunningQueryState = {
+    panelData: {
+      state: LoadingState.Loading,
+      series: [],
+      request: request,
+    },
+    packets: {},
+  };
+
+  // Return early if there are no queries to run
+  if (!request.targets.length) {
+    request.endTime = Date.now();
+    state.panelData.state = LoadingState.Done;
+    return of(state.panelData);
+  }
+
+  const dataObservable = callQueryMethod(datasource, request).pipe(
+    // Transform response packets into PanelData with merged results
+    map((packet: DataQueryResponse) => {
+      if (!isArray(packet.data)) {
+        throw new Error(`Expected response data to be array, got ${typeof packet.data}.`);
+      }
+
+      request.endTime = Date.now();
+
+      state = processResponsePacket(packet, state);
+      return state.panelData;
+    }),
+    // handle errors
+    catchError(err =>
+      of({
+        ...state.panelData,
+        state: LoadingState.Error,
+        error: processQueryError(err),
+      })
+    ),
+    // finalize is triggered subscriber unsubscribes
+    // This makes sure any still running network requests are cancelled
+    finalize(cancelNetworkRequestsOnUnsubscribe(request)),
+    // this makes it possible to share this observable in takeUntil
+    share()
+  );
+
+  // If 50ms without a response emit a loading state
+  // mapTo will translate the timer event into state.panelData (which has state set to loading)
+  // takeUntil will cancel the timer emit when first response packet is received on the dataObservable
+  return merge(
+    timer(50).pipe(
```

Comment:
Would it make sense to increase the interval? I assume it's only the first call of this interval that really matters otherwise it will emit the same value all the time until the data is loaded.

Reply:
Ah wait, it actually emits only once as it has one arg provided, sorry!